### DISCUSSION TITLE
DeepHomoModel resnet18 architecture

### DISCUSSION THREAD

**Discussion by karlosos** [DonsetPG/narya#15] [Q&A] [answered]
I'm trying to recreate the deep homography model from scratch and experiment with different approaches. DeepHomoModel, which is implemented in [models/keras_models.py], uses `resnet18` as a backend if I understand correctly. The `resnet18` model is built from a .h5 file that is downloaded from https://storage.googleapis.com/narya-bucket-1/models/deep_homo_model_1.h5.

What is original source of `deep_homo_model_1.h5`? Are there any resources on how to build it or what is the original source of this model? I found out that you can download weights from various sources, e.g. https://github.com/qubvel/classification_models/releases/tag/0.0.1. Are those the weights (architecture) that were used initially?

EDIT 1: Ok, I've tried using weights downloaded from https://github.com/qubvel/classification_models/releases/tag/0.0.1 and they are not working. That's because they are only weights not the whole model and I get the error `ValueError: No model found in config file.`.

[models/keras_models.py]: https://github.com/DonsetPG/narya/blob/master/narya/models/keras_models.py

**Comment by DonsetPG** [comment, marked as answer]
Hey,

> DeepHomoModel, which is implemented in [models/keras_models.py](https://github.com/DonsetPG/narya/blob/master/narya/models/keras_models.py), uses resnet18 as a backend if I understand correctly.

Yes, that's right.

The DeepHomoModel story is quite complicated, haha. I wanted to reuse the method from [Optimizing Through Learned Errors for Accurate Sports Field Registration](https://openaccess.thecvf.com/content_WACV_2020/papers/Jiang_Optimizing_Through_Learned_Errors_for_Accurate_Sports_Field_Registration_WACV_2020_paper.pdf), since the results were great, and they even provided [codes and trained models](https://github.com/vcg-uvic/sportsfield_release) (which is amazing for this field). 

However, they did filed a patent on their method, thus making any commercial use impossible, and even if I was not planning on doing so, I was not comfortable enough to use their method while it was patented. 

I figure that I could still use the architecture of the first model they used, and this is where the file https://storage.googleapis.com/narya-bucket-1/models/deep_homo_model_1.h5 comes from (I used a library to translate the torch model into a Keras model, as I prefer to use Keras/TF instead of Torch. I wanted to do the same thing for the Player Re-Identification model but could not make it work, sadly).

So that's the story of the DeepHomoModel that is developed here. 
To be completely honest, this is a super tricky model to train, ask @ArthurVerrez but I think I iterated for over a month on this model during the first lockdown. If you can, I would try to first train a keypoint model. I also found this paper a few weeks ago that might be interesting: [A Robust and Efficient Framework for Sports-Field Registration](https://openaccess.thecvf.com/content/WACV2021/papers/Nie_A_Robust_and_Efficient_Framework_for_Sports-Field_Registration_WACV_2021_paper.pdf)

I hope this is useful.

**Comment by Abhipray04** [comment]
Hey @DonsetPG and @karlosos, I am trying to train the DeepHomoModel and finding it tricky to train. Is there a flowchart or steps to follow to ensure I am correctly training the model? Also, I am trying to train the model on custom data.

**Comment by DonsetPG** [comment]
Hey, you can have a look at this file https://github.com/DonsetPG/narya/blob/master/narya/trainer/homography_train.py for the training. It should work on a custom dataset.
